RecoCTPPS/ProtonReconstruction/src/ProtonReconstructionAlgorithm.cc

```diff
@@ -0,0 +1,391 @@
+
+void ProtonReconstructionAlgorithm::release()
+{
+  initialized_ = false;
+
+  m_rp_optics_.clear();
+}
+
+//----------------------------------------------------------------------------------------------------
+
+double ProtonReconstructionAlgorithm::ChiSquareCalculator::operator() (const double* parameters) const
+{
+  // extract proton parameters
+  const LHCOpticalFunctionsSet::Kinematics k_in = { 0., parameters[1], parameters[3], parameters[2], parameters[0] };
+
+  // calculate chi^2 by looping over hits
+  double S2 = 0.;
+
+  for (const auto &track : *tracks) {
+    const CTPPSDetId rpId(track->getRPId());
+
+    // transport proton to the RP
+    auto oit = m_rp_optics->find(rpId);
+    LHCOpticalFunctionsSet::Kinematics k_out;
+    oit->second.optics->transport(k_in, k_out);
+
+    // proton position wrt. beam
+    const double& x = k_out.x - oit->second.x0;
+    const double& y = k_out.y - oit->second.y0;
+
+    // make sure that uncertainties are reasonable
+    double x_unc = track->getXUnc();
+    if (x_unc < 1E-3)
+      x_unc = 40E-3;
+
+    double y_unc = track->getYUnc();
+    if (y_unc < 1E-3)
+      y_unc = 40E-3;
+
+    // calculate chi^2 contributions, convert track data mm --> m
+    const double x_diff_norm = (x - track->getX()*1E-3) / (x_unc*1E-3);
+    const double y_diff_norm = (y - track->getY()*1E-3) / (y_unc*1E-3);
+
+    // increase chi^2
+    S2 += x_diff_norm*x_diff_norm + y_diff_norm*y_diff_norm;
+  }
+
+  return S2;
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void ProtonReconstructionAlgorithm::reconstructFromMultiRP(
+  const reco::ProtonTrackExtra::CTPPSLocalTrackLiteRefVector &tracks,
+  std::vector<reco::ProtonTrack> &output,
+  std::vector<reco::ProtonTrackExtra> &outputExtra,
+  const LHCInfo &lhcInfo, std::ostream& os) const
+{
+  if (!initialized_)
+    return;
+
+  // need at least two tracks
+  if (tracks.size() < 2)
+    return;
+
+  // make sure optics is available for all tracks
+  for (const auto &it : tracks)
+  {
+    auto oit = m_rp_optics_.find(it->getRPId());
+    if (oit == m_rp_optics_.end())
+      throw cms::Exception("ProtonReconstructionAlgorithm") << "Optics data not available for RP " <<
+        it->getRPId() << ", i.e. " << CTPPSDetId(it->getRPId()) << ".";
+  }
+
+  // initial estimate of xi and th_x
+  double xi_init = 0., th_x_init = 0.;
+
+  const bool use_improved_estimate = true;
+  if (use_improved_estimate)
+  {
+    double x_N = 0., x_F = 0.;
+    const RPOpticsData *i_N = nullptr, *i_F = nullptr;
+    unsigned int idx = 0;
+    for (const auto &track : tracks)
+    {
+      auto oit = m_rp_optics_.find(track->getRPId());
+
+      if (idx == 0) { x_N = track->getX() * 1E-3; i_N = &oit->second; }
+      if (idx == 1) { x_F = track->getX() * 1E-3; i_F = &oit->second; }
+      if (idx == 2) break;
+
+      idx++;
+    }
+
+    const double a = i_F->ch1*i_N->la1 - i_N->ch1*i_F->la1;
```

Comment:
this is flagged by the static analyzer
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25495/32272/llvm-analysis/report-e469ca.html#EndPath

```c++
x_N = tracks[0].getX()*1E-3; i_N = &m_rp_optics_.at(tracks[0].getRPId()).second;
x_F = tracks[1].getX()*1E-3; i_F = &m_rp_optics_.at(tracks[1].getRPId()).second;
```
is apparently equivalent and should clear the static analysis issues.

Reply:
Done.